syncserver/src/server/user_agent.rs

```diff
@@ -11,6 +14,170 @@ const VALID_UA_BROWSER: &[&str] = &["Chrome", "Firefox", "Safari", "Opera"];
 // field). Windows has many values and we only care that its Windows
 const VALID_UA_OS: &[&str] = &["Firefox OS", "Linux", "Mac OSX"];
 
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub enum Platform {
+    FirefoxDesktop,
+    Fenix,
+    FirefoxIOS,
+    Other,
+}
+
+impl fmt::Display for Platform {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let name = format!("{:?}", self).to_lowercase();
+        write!(fmt, "{}", name)
+    }
+}
+
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub enum DeviceFamily {
+    Desktop,
+    Mobile,
+    Tablet,
+    Other,
+}
+
+impl fmt::Display for DeviceFamily {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let name = format!("{:?}", self).to_lowercase();
+        write!(fmt, "{}", name)
+    }
+}
+
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub enum OsFamily {
+    Windows,
+    MacOs,
+    Linux,
+    IOS,
+    Android,
+    Other,
+}
+
+impl fmt::Display for OsFamily {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let name = format!("{:?}", self).to_lowercase();
+        write!(fmt, "{}", name)
+    }
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct DeviceInfo {
+    pub platform: Platform,
+    pub device_family: DeviceFamily,
+    pub os_family: OsFamily,
+    pub firefox_version: u32,
+}
+
+impl DeviceInfo {
+    /// Determine if the device is a desktop device based on either the form factor or OS.
+    pub fn is_desktop(&self) -> bool {
+        matches!(&self.device_family, DeviceFamily::Desktop)
+            || matches!(
+                &self.os_family,
+                OsFamily::MacOs | OsFamily::Windows | OsFamily::Linux
+            )
+    }
+
+    /// Determine if the device is a mobile phone based on either the form factor or OS.
+    pub fn is_mobile(&self) -> bool {
+        matches!(&self.device_family, DeviceFamily::Mobile)
+            && matches!(&self.os_family, OsFamily::Android | OsFamily::IOS)
+    }
+
+    /// Determine if the device is iOS based on either the form factor or OS.
+    pub fn is_ios(&self) -> bool {
+        matches!(&self.device_family, DeviceFamily::Mobile)
+            && matches!(&self.os_family, OsFamily::IOS)
+    }
+
+    /// Determine if the device is an android (Fenix) device based on either the form factor or OS.
+    pub fn is_fenix(&self) -> bool {
+        matches!(&self.device_family, DeviceFamily::Mobile)
+            && matches!(&self.os_family, OsFamily::Android)
+    }
+}
+
+pub fn get_device_info(user_agent: &str) -> DeviceInfo {
+    let mut w_result = Parser::new()
+        .parse(user_agent)
+        .unwrap_or_else(|| WootheeResult {
```

Comment:
WootheeResult implements `Default` so this can be a lot easier.
```suggestion
        .unwrap_or_default();
```

Reply:
Great suggestion, thx 👍